shopping-cart/shopping-cart-scala/shopping-cart/src/main/scala/com/example/shoppingcart/impl/ShoppingCart.scala

```diff
@@ -0,0 +1,250 @@
+   * means the state gets stored to the database, so that when the entity gets
+   * loaded, you don't need to replay all the events, just the ones since the
+   * snapshot. Hence, a JSON format needs to be declared so that it can be
+   * serialized and deserialized when storing to and from the database.
+   */
+  implicit val format: Format[ShoppingCart] = Json.format
+}
+
+final case class ShoppingCart(items: Map[String, Int], checkedOut: Boolean) {
+
+  import ShoppingCart._
+
+  //The shopping cart behavior changes if it's checked out or not. The command handles are different for each case.
+  def applyCommand(cmd: Command): ReplyEffect[Event, ShoppingCart] =
+    if (checkedOut) {
+      cmd match {
+        // it is allowed to read it's state of a checked-out cart
+        case Get(replyTo) => Effect.reply(replyTo)(toSummary(this))
+        // CheckedOut is a final state, no mutations allowed
+        case AddItem(_, _, replyTo) => Effect.reply(replyTo)(Rejected("Cannot add an item to a checked-out cart"))
+        case RemoveItem(_, replyTo) => Effect.reply(replyTo)(Rejected("Cannot remove an item from a checked-out cart"))
+        case AdjustItemQuantity(_, _, replyTo) =>
+          Effect.reply(replyTo)(Rejected("Cannot adjust an item quantity on a checked-out cart"))
+        case Checkout(replyTo) => Effect.reply(replyTo)(Rejected("Cannot checkout a checked-out cart"))
+      }
+    } else {
+      cmd match {
+        case AddItem(itemId, quantity, replyTo)            => addItem(itemId, quantity, replyTo)
+        case RemoveItem(itemId, replyTo)                   => removeItem(itemId, replyTo)
+        case AdjustItemQuantity(itemId, quantity, replyTo) => adjustItemQuantity(itemId, quantity, replyTo)
+        case Checkout(replyTo)                             => checkout(replyTo)
+        case Get(replyTo)                                  => Effect.reply(replyTo)(toSummary(this))
+      }
+    }
+
+  private def checkout(replyTo: ActorRef[Confirmation]): ReplyEffect[Event, ShoppingCart] = {
+    if (items.isEmpty)
+      Effect.reply(replyTo)(Rejected("Cannot checkout an empty shopping cart"))
+    else
+      Effect
+        .persist(CartCheckedOut(Instant.now()))
+        .thenReply(replyTo)(updatedCart => Accepted(toSummary(updatedCart)))
+  }
+
+  private def addItem(
+      itemId: String,
+      quantity: Int,
+      replyTo: ActorRef[Confirmation]
+  ): ReplyEffect[Event, ShoppingCart] = {
+    if (items.contains(itemId))
+      Effect.reply(replyTo)(Rejected(s"Item '$itemId' was already added to this shopping cart"))
+    else if (quantity <= 0)
+      Effect.reply(replyTo)(Rejected("Quantity must be greater than zero"))
+    else
+      Effect
+        .persist(ItemAdded(itemId, quantity))
+        .thenReply(replyTo)(updatedCart => Accepted(toSummary(updatedCart)))
+  }
+
+  private def removeItem(itemId: String, replyTo: ActorRef[Confirmation]): ReplyEffect[Event, ShoppingCart] = {
```

Comment:
We should reconsider the name `removeItem`. It is overloaded:

```scala
  private def removeItem(String, ActorRef[...]): ReplyEffect[...] = ???
  private def removeItem(String): ShoppingCart = ???
```

the problem is one of the overloads is a command handler and the other is an event handler. I find this terribly confusing. How about:

```scala
  private def removeItem(String, ActorRef[...]): ReplyEffect[...] = ???
  private def onItemRemoved(String): ShoppingCart = ???
```
So command handler methods use an imperative (probably the same exact form used by the actual command, in this case `class RemoveItem`). But event handlers use the `onEventName` form: prefix `on` followed by the event name.

Finally, in this sample there's a particular case:

```scala
      case ItemAdded(itemId, quantity)            => addOrUpdateItem(itemId, quantity)
      case ItemQuantityAdjusted(itemId, quantity) => addOrUpdateItem(itemId, quantity)
...
  private def addOrUpdateItem(...)
```

which I'd rename to:

```scala
private def onItemAddedOrUpdated(...)
```
which is not exactly "prefix `on` followed by the event name" but is clear enough.